packages/shylu/shylu_dd/bddc/src/shylu_DisconnectedComponentChecker.h

```diff
@@ -0,0 +1,418 @@
+  }
+
+ private: // variables
+  LO m_numSubNew{0}, m_numMatNew{0};
+  LO **m_subRowBeginPtr{0}, **m_subColumnsPtr{0};
+  SX **m_subValuesPtr{0};
+  std::vector< std::vector<LO> > m_subNodesNew, m_subRowBeginNew,
+    m_subColumnsNew;
+  std::vector< std::vector<SX> > m_subValuesNew;
+
+ private: // functions
+
+  void getColNodes(const LO numNodes, 
+		   const LO* nodeBegin, 
+		   std::vector<LO> & colNodes)
+  {
+    for (LO i=0; i<numNodes; i++) {
+      for (LO j=nodeBegin[i]; j<nodeBegin[i+1]; j++) {
+	colNodes[j] = i;
+      }
+    }
+  }
+
+  void getSubNodeBegin(const std::vector< std::vector<LO> > & subNodes, 
+		       const LO* nodeBegin, 
+		       std::vector< std::vector<LO> > & subNodeBegin)
+  {
+    const LO numSub = subNodes.size();
+    subNodeBegin.resize(numSub);
+    for (LO j=0; j<numSub; j++) {
+      const LO numNodes = subNodes[j].size();
+      subNodeBegin[j].resize(numNodes+1, 0);
+      for (LO i=0; i<numNodes; i++) {
+	const LO node = subNodes[j][i];
+	const LO numDofNode = nodeBegin[node+1] - nodeBegin[node];
+	subNodeBegin[j][i+1] = subNodeBegin[j][i] + numDofNode;
+      }
+    }
+  }
+
+  void identifyComponents(const LO numNodes,
+			  const LO* nodeBegin,
+			  const std::vector< std::vector<LO> > & subNodes,
+			  const std::vector< std::vector<LO> > & subNodeBegin,
+			  LO** subRowBegin,
+			  LO** subColumns,
+			  SX** subValues,
+			  std::vector< std::vector<int> > & subNodeComponents)
+  {
+    // Note: we place nodes without any degrees of freedom or not
+    //       connected to any other nodes in component 0
+    const LO numSub = subNodes.size();
+    const LO numDof = nodeBegin[numNodes];
+    std::vector<LO> nodeMap(numNodes, -1), activeNodes(numNodes),
+      colNodes(numDof);
+    std::vector<bool> nodeFlag(numNodes, false);
+    subNodeComponents.resize(numSub);
+    for (LO i=0; i<numSub; i++) {
+      std::vector<int> nodeConn, nodeConnPtr;
+      const LO numNodes = subNodes[i].size();
+      determineNodalConnectivity
+	(numNodes, subNodeBegin[i].data(), subRowBegin[i], subColumns[i],
+	 activeNodes, colNodes, nodeFlag, nodeConn, nodeConnPtr);
+      // make adjustments to nodeConn and nodeConnPtr to not include 
+      // any disconnected nodes
+      LO numConnected(0);
+      for (LO j=0; j<numNodes; j++) {
+	const LO numAdj = nodeConnPtr[j+1] - nodeConnPtr[j];
+	if (numAdj > 0) nodeMap[j] = numConnected++;
+      }
+      std::vector<int> nodeConnPtrNew(numConnected+1, 0);
+      numConnected = 0;
+      LO numTerms(0);
+      for (LO j=0; j<numNodes; j++) {
+	const LO numAdj = nodeConnPtr[j+1] - nodeConnPtr[j];
+	if (numAdj > 0) {
+	  for (LO k=nodeConnPtr[j]; k<nodeConnPtr[j+1]; k++) {
+	    const LO node = nodeMap[nodeConn[k]];
+	    BDDC_TEST_FOR_EXCEPTION(node == -1, std::runtime_error, 
+				    "invalid node number");
+	    nodeConn[k] = node;
+	  }
+	  numTerms += numAdj;
+	  nodeConnPtrNew[++numConnected] = numTerms;
+	}
+      }
+      int *components(0);
```

Comment:
Please initialize pointers with `nullptr`, not with `0`.  More importantly,  consider using `std::unique_ptr<int[]>` instead of a raw pointer here.

Reply:
Done. I had trouble converting to std::unique_ptr<int[]> for this case. I need to understand how to use std::unique_ptr.

Reply:
We discussed this particular example over the phone today.